Updating thp when valid VFP table is specified

GitPaean opened the pull request:
With the old implementation, VFP table number is only parsed when there is a THP constraint/target is specified. 

While we need to update/calculate the THP when the well has a valid VFP table associated with it for outputting purpose. 

When there is no valid VFP table specified, we set the thp value to be zero. 

It will make the model 2 history period able to output thp values for comparison. 

This is a PR should only impact output when THP constraint/target is not present, while some rates can result in problematic ratios causing the VFP/THP calculation failure.

GitPaean committed:
adding function isVFPActive() for WellInterface

Even the well does not have a THP target/constraint, but if it is
specified with a valid VFP table, we are supposed to update the thp
value for output purposes.

GitPaean committed:
updating the thp value if VFP table is valid

through function isVFPActive().

GitPaean commented:
jenkins build this please

GitPaean commented:
It does not change anything of the simulation when tested with Norne and Model 2. 

It also make some of the wells in model 2 output WTHP for comparison. 

I would like to add a few more guard and debugging output in case there is some `NAN` or `INF` happens during the simulation related to future cases we can not foresee to avoid sudden termination of the simulation without sensible message.

GitPaean committed:
detecting the NAN and INF values during VFP calculation

and give warning messages when they happen, since they can be the
culprits of the termination of simulation later.

For ADB related, I did not find a good way to do the detection, so there
is no warning message given.

GitPaean commented:
jenkins build this please

atgeirr commented:
I think we will merge this after the VFP split PR.

GitPaean commented:
> I think we will merge this after the VFP split PR.

Sure. It is fine with me. 

However, depending how far you have been with the VFP splitting process, it might not be a bad idea to merge this one first. Since this PR does not change the splitting process much.

atgeirr commented:
This should now be updated, as the splitting PR has been merged.

atgeirr commented:
jenkins build this please

atgeirr commented on opm/autodiff/WellInterface_impl.hpp:
```diff
@@ -979,6 +979,50 @@ namespace Opm
+            if (table_id <= 0) {
+                return false;
+            } else {
+                if (vfp_properties_->getProd()->getTable(table_id)) {
```
Since getTable() will throw if invalid id, the else branch will never be reached. Just return true is fine here (and for injector).

atgeirr reviewed (changes requested):
Looks fine, just a minor adjustment needed.

GitPaean committed:
addressing review comments from PR 1622

GitPaean commented:
jenkins build this please

GitPaean commented:
jenkins build this please

GitPaean committed:
remove a useless #if HAVE_OPENMP in BlackoilWellModel

GitPaean commented:
I removed a useless `HAVE_OPENMP` in BlackWellMode_impl.hpp

It is just too tiny to create a new PR for this. Please let me know if it is a problem.

atgeirr commented on opm/autodiff/WellInterface_impl.hpp:
```diff
@@ -979,6 +979,42 @@ namespace Opm
+            if (table_id <= 0) {
+                return false;
+            } else {
+                if (vfp_properties_->getInj()->getTable(table_id)) {
```
I think this `if` should also be removed, the code should just be:
```
return table_id > 0;
```
(also in production part).

GitPaean replied:
Then we will miss the `throw` .

GitPaean replied:
Or, we get a `throw` in the later calculation.  I do not mind either and I do not think the current form is bad.

atgeirr replied:
Indeed, it will throw from somewhere else. But calling a method to see if it throws is not a good approach. If we want to dig deeper here (I do not think we should) we should add a haveTable() method we could call, and make a nice message here, but for the current use, it would be overengineering.

GitPaean replied:
I think this is a nice point. Let me do that. It should be quick.

GitPaean committed:
adding hasTable() method to the VFP properties

and give a clear message about non-accessible table ID in the function
isVFPActive() in WellInterface.

GitPaean commented:
jenkins build this please
